File: app/src/main/java/me/yxp/qfun/hook/api/OnGetMsgRecord.java
package me.yxp.qfun.hook.api;

import java.lang.reflect.Method;

import me.yxp.qfun.hook.base.ApiHookItem;
import me.yxp.qfun.utils.hook.HookUtils;
import me.yxp.qfun.utils.reflect.ClassUtils;
import me.yxp.qfun.utils.reflect.MethodUtils;

public final class OnGetMsgRecord extends ApiHookItem {
    public static final ApiHookItem INSTANCE = new OnGetMsgRecord();

    @Override
    public void loadHook() throws Throwable {
        Method getMsgRecord = MethodUtils.create(ClassUtils._AIOMsgItem())
                .withMethodName("getMsgRecord")
                .findOne();

        HookUtils.hookAlways(getMsgRecord, null, param -> {
            Object msgRecord = param.getResult();
            for (Listener listener : mListenerSet) {
                if (listener instanceof GetMsgRecordListener) {
                    ((GetMsgRecordListener) listener).onGet(msgRecord);
                }
            }
        });
    }

    public interface GetMsgRecordListener extends Listener {
        void onGet(Object msgRecord) throws Throwable;
    }
}

File: app/src/main/java/me/yxp/qfun/hook/api/OnGetMsgRecord.kt
package me.yxp.qfun.hook.api

import com.tencent.mobileqq.aio.msg.AIOMsgItem
import com.tencent.qqnt.kernel.nativeinterface.MsgRecord
import me.yxp.qfun.annotation.HookItemAnnotation
import me.yxp.qfun.hook.base.BaseApiHookItem
import me.yxp.qfun.hook.base.Listener
import me.yxp.qfun.utils.hook.hookAfter

@HookItemAnnotation("监听获取消息记录")
object OnGetMsgRecord : BaseApiHookItem<GetMsgRecordListener>() {

    override fun loadHook() {
        AIOMsgItem::class.java.getDeclaredMethod("getMsgRecord")
            .hookAfter(this) { param ->
                val msgRecord = param.result as MsgRecord
                listenerSet.filter { verify(it) }
                    .forEach { it.onGet(msgRecord) }
            }
    }
}

fun interface GetMsgRecordListener : Listener {
    fun onGet(msgRecord: MsgRecord)
}